SwiftCompilerSources/Sources/Optimizer/FunctionPasses/DiagnoseLifetimeDependence.swift

```diff
@@ -0,0 +1,570 @@
+///  -> %nonEscapableValue = copy_value %inner : $NonEscapableStruct
+///     %_ = apply @usesNonEscapable(%nonEscapableValue)
+///          : $(NonEscapableStruct) -> ()
+///     end_access %access : $*Container
+///
+/// Definition: lifetime introducer. Any owned or guaranteed value that is not produced by a forwarding operation or a copy. (`begin_borrow` of an owned value is non-forwarding so it does introduce a lifetime).
+///
+/// Definition: non-escaping scope. The lifetime of value that introduces the lifetime of a non-escapable value.
+///
+/// Each non-escapable value in a function has a forwarded and copied lifetime introducer. The lifetime of %nonEscapableValue is introdued by %inner.
+///
+///     %inner = apply @forwardsNonEscapable(%outer)
+///               : $(@resultDependsOn @owned NonEscapableStruct)
+///
+/// The lifetime introducer may be recognized as an instruction that transfers lifetime dependence from one value to another. An apply with a @resultDependsOn parameter transfers lifetime dependence from the parameter (the parent lifetime) to the result (the dependent lifetime).
+///
+/// Lifetime *propagation* occurs when the parent lifetime is itself nonescapable. Finding the non-escaping scope for lifetime-propagation involves recursively asking for the parent's introducer. In this example, recursion gives us %outer as the next introducer, and %access as the next parent lifetime:
+///
+///     %outer = apply @returnsNonEscapable(%access)
+///               : $(@resultDependsOn @in Container) -> NonEscapableStruct
+///
+/// Lifetime *encapsulation* (as opposed to propagation) occurs when the parent lifetime is escapable. In this case, we find the parent's enclosing scope as follows:
+///
+/// For owned values, the value's lifetime is itself the enclosing scope, delimited by the consumes after following any forwarding operations.
+///
+/// For borrowed values, the enclosing scope is created by the borrow introducer (visitBorrowIntroducers).
+///
+/// For addresses, the enclosing scope is created by an access scope (enclosingAcessScope). Note that an exclusive access scope *must* exist for all non-escaping values that depend on an addressible value.
+///
+/// In this example, the parent value is an address, so the non-escaping scope is an access scope:
+///
+///     %access = begin_access %address : $*Container
+///
+//===----------------------------------------------------------------------===//
+
+import SIL
+
+// DiagnosticEngine already declares DiagID. Why isn't it visible?
+import ASTBridging
+public typealias DiagID = BridgedDiagID
+
+private let verbose = true
+
+private func log(_ message: @autoclosure () -> String) {
+  if verbose {
+    print("### \(message())")
+  }
+}
+
+extension InstructionRange {
+  // TODO: InstructionRange strangely does not support phis or function args. This is problematic, because the block may be mutated.
+  static func getRepresentativeInstruction(value: Value) -> Instruction {
```

Comment:
Is this function still needed?

Reply:
`getRepresentativeInstruction` is needed. The use wasn't pushed. I didn't create a nice initializer for this because it's a terrible hack that people shouldn't use accidentally.